### PR TITLE
Refactor All The Things

### PR DESCRIPTION
**Pull request by gordonwritescode**
This PR includes a pretty massive refactor of the project to implement some best practices, improve maintainability, remove cruft, and optimize performance.

* [x] Remove statically included dependencies like Ladda, jQuery, etc
* [x] Remove RequireJS in favor of using CommonJS everywhere
* [x] Add `.jshintrc` to enforce conventions and code style
* [x] Refactor `userdata.js`, `updater.js`, `dataserv.js`, `logger.js` to use an OOP approach
* [x] Untangle HTML and JS code
* [x] Add Vue.js to clean up and abstract UI behavior code
* [x] More sane directory/project structure
* [x] Make use of JSDoc comments for *everything*
* [x] Remove transpilation task and other extraneous nonsense from boilerplate
* [x] Re-implement validation of user data in main view
* [x] Re-implement configuration commission to disk in main view
* [x] Plug in new `DataServWrapper` class to the main view
* [x] Fix lack of visual feedback on farming start (re-implement "ladda" button)
* [x] Refactor `setup.js` script to `DataServInstaller` class
* [x] Wire up `DataServInstaller` to UI
* [x] Author README, include style guide, and architecture overview
* [x] Add config migration to ensure that old configs are re-structured
* [x] Log view should separate logs between different drives

---

Configuration files from previous versions will automatically be converted to the new format.

### CODE REVIEW

**Comment by gordonwritescode** [comment]
I will attempt to resolve the merge conflicts tomorrow.

**Comment by gordonwritescode** [comment]
Getting close, should be ready for review tomorrow.

**Comment by gordonwritescode** [comment]
This also closes #66 #64 and #61

**Comment by gordonwritescode** [comment]
@super3 @gamedevsam this still needs to be tested on Windows and OSX but can be reviewed preemptively.

**Comment by littleskunk** [comment]
Test with windows 7 64bit (installed 32bit node.js)
git clone https://github.com/gordonwritescode/driveshare-gui.git -b refactor/allthethings

Browser button dont work.
Start and running botton are both visible at the same time.
Start button dont work.

**Comment by gordonwritescode** [comment]
Okay, I've got everything running on Windows 7 64bit now - with the exception of https://github.com/Storj/driveshare-gui/issues/69

UI render time seems to suffer on windows - gnu/linux is very fast - perhaps it's my VM. Moving on to OSX tomorrow.

**Comment by littleskunk** [comment]
Same problem with the last commit. No button is working for me and i have that exception in my npm log.

**Comment by gordonwritescode** [comment]
You need to remove your old config file. That line is expecting a different structure:

https://github.com/gordonwritescode/driveshare-gui/blob/refactor/allthethings/app/lib/views.js#L288

I'll be sure to write a config "migration" piece before this gets pushed to users.

**Comment by littleskunk** [comment]
Yes that fixed it for me. Now everything is working with windows. Well done. I get a message box with a nice dataserv-client exception and was able to fix the problem in in a minute.

**Comment by gordonwritescode** [comment]
Now closes #70

**Comment by gordonwritescode** [comment]
Working on virtualizing OSX for surface testing this PR, then I think we are good for merge. Will update shortly.

**Comment by gordonwritescode** [comment]
Okay, my Hackintosh VM is DOA - will need to borrow someone's Mac for about 20 minutes. @frdwrd

**Comment by littleskunk** [comment]
I can give you a Mac OS X Yosemite vmdk image

**Comment by gordonwritescode** [comment]
@littleskunk that would be *super*!

**Comment by littleskunk** [comment]
whats your user name on slack? dont want to write the link here.

**Comment by gordonwritescode** [comment]
`gordon` :)

**Comment by gordonwritescode** [comment]
Closes #71

**Comment by gordonwritescode** [comment]
Now closes #74

**Comment by gordonwritescode** [comment]
This PR is now merged with the [develop](https://github.com/Storj/driveshare-gui/tree/develop) branch. Deferring merge with `master` until Monday. Cheers!

**Comment by gordonwritescode** [comment]
Closes #73

**Comment by gordonwritescode** [comment]
Now includes fixes for #75 and #64 - updated on `develop`.

**Comment by gamedevsam** [comment]
Can you make this PR into the dev branch? It would be good to make the refactor there so we can test it thoroughly before merging it into master.

**Comment by gordonwritescode** [comment]
@gamedevsam this PR is already mirrored on `develop`

**Comment by gamedevsam** [comment]
Reviewing the code now, will merge unless something glaring jumps out at me.

**Comment by super3** [comment]
@gamedevsam waiting till Monday to merge

File: app/lib/installer.js [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+ * Initializes the installer and begins emitting events
+ * #install
+ * @param {String} password - gnu/linux only
+ */
+DataServInstaller.prototype.install = function(password) {
+  if (Object.keys(this._targets).indexOf(this._platform) === -1) {
+    return this.emit('error', new Error('This platform is not supported'));
+  }
+
+  var self = this;
```
Comment (gamedevsam) [inline app/lib/installer.js]: Seems like self and this are used somewhat interchangeably in this file. Seems a bit odd, is this done on purpose?

Comment (gordonwritescode) [inline app/lib/installer.js]: The value of `this` may change inside closures, so it's a good practice to store a reference to the parent scope for use inside callbacks that execute within a separate context.

Comment (gamedevsam) [inline app/lib/installer.js]: :+1: 

Reviewing this code almost brought a tear to my eyes. This is a much more coherent way of structuring the project, lots of improvements, very happy!

Comment (gamedevsam) [inline app/lib/installer.js]: @super3: I'm deferring merging duties to you and @gordonwritescode until further notice.

**Comment by gordonwritescode** [comment]
Now closes #63

**Comment by gordonwritescode** [comment]
Tested on OSX - :+1:

**Comment by gordonwritescode** [comment]
@super3 :shipit:

**Comment by super3** [comment]
@gordonwritescode Merging without tests. What has this world come to?

**Comment by gordonwritescode** [comment]
@super3 tests are on there way: https://github.com/Storj/driveshare-gui/pull/78

**Pull request merged**